### ISSUE TITLE
Feature suggestion: get list of tests and run them manually

### ISSUE THREAD

**Issue by andreaplanet** [onqtam/doctest#184]
If I'm not wrong, the tests are executed in a sort of random order (based on linkage), so I cannot specify a dependency between tests or an execution order. For example a low level math test suite should be executed before a high level suite that uses the low level math functions.

A solution could be the possibility to get the list of tests to execute (based on command line arguments) and then the ability to call individual tests on my own. I could handle the execution order on my own and so even run tests in parallel #168 using separate threads. And if I replace the memory allocators I could implement #80.

And if the test result is returned as a structure instead of printing it then I may could write my own reporter, see #138, #181 

Well it's just an idea...

**Comment by onqtam** [comment]
Hi! You could use the ```--order-by=<type>``` [command line parameter](https://github.com/onqtam/doctest/blob/master/doc/markdown/commandline.md) - by default test cases are ordered by the file they are defined in, and within a single file - based on the line they are written at. You could either put prefixes in the names of test cases such as ```TEST_CASE("[prefix] testing x")``` and then use ```--order-by=name``` in which case tests will be ran after sorted lexicographically (so you would have to think about the lexicographical order when naming), and you could also use [test suites](https://github.com/onqtam/doctest/blob/master/doc/markdown/testcases.md#test-suites) and use ```--order-by=suite```.

You could also not think about the lexicographical sorting and instead provide [your own main](https://github.com/onqtam/doctest/blob/master/doc/markdown/main.md) and use the [filters](https://github.com/onqtam/doctest/blob/master/doc/markdown/commandline.md) such as ```--test-case=<wildcard>``` and ```--test-case-exclude=<wildcard>``` (or ```test-suite``` and ```test-suite-exclude```) like this:

```c++
context.addFilter("test-case", "*math*");
context.run(); // run only math-y tests
context.clearFilters();
context.addFilter("test-case-exclude", "*math*");
context.run(); // run anything but math-y tests
```

I acknowledge there is a need for an API for getting the registered tests and perhaps running them, but that would be a bit complicated and is low priority.

For the parallel execution of test cases - it wouldn't be trivial to get [```SUBCASE```s](https://github.com/onqtam/doctest/blob/master/doc/markdown/tutorial.md#test-cases-and-subcases) and [logging functionality]() to work in that case - but at least you could invoke the binary with the tests multiple times with different ranges, etc.

**Comment by onqtam** [comment]
Do you have interest in a [decorator](https://github.com/onqtam/doctest/blob/master/doc/markdown/testcases.md#decorators) for specifying the order/priority - like this:

```c++
TEST_CASE("name" * doctest::priority(10)) {
    // ...
}
```

**Comment by andreaplanet** [comment]
When all tests succeed then the order of execution is basically irrelevant, it affects the output order. It matters when a test fails and so many subsequent tests will fail too in different suites and you have to find out who originated it. 

As you suggested, within a Test Suite I sort the code of the Test Cases within the cpp file by the desired execution order (line). Placing basic tests at the beginning and the most complex/dependent tests at the end. No need for more specifications (decorator) within a Suite, e.g. for individual tests.

I still don't know which could be the best solution for specifying dependencies. For example if a math test suite fails one of it's test then I don't want to execute a higher test suite based on the math stuff because it's precondition isn't satisfied (a working math). With a priority decorator on the test suites I only sort the execution order and I have to self organize a class of priority numbers. It would be easier in the main function to specify only direct dependencies between test suites, even just with a single text line "Cube>Math, Math>Number, ImageTransformations>Image" for all suites (better overview of the dependencies) or a text decorator (more elegant) for the suite, like TEST_SUITE_BEGIN("Cube" *  doctest::dependsOnSuite("Math")). Which sound like "don't run Cube if Math fails".
I don't know if for a general use this could have sense.

Personally at the moment I only have 30 test suites and I can manage that without this feature. 
So right now this feature is not a priority at all... more something to think about.
I will let you know in the future, once the suites will grow in number.

Thank you :-)

**Comment by cdeln** [comment]
I think we can close this since there are several workarounds.

Implementing a dependency graph solver feels overkill and out of scope for this project.

A small suggestion: if your project is becoming very complex with many different modules depending on each other, then that is a call for refactoring I would say. Consider breaking out some lower level code into a different project/folder. Then you can run all tests related to that part of your code base in isolation, and you can control the order of test execution in your build system instead.
